src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -190,6 +201,68 @@ balanceTxWithConstraints unbalancedTx constraintsBuilder = do
       =<< maybe (liftQueryM QueryM.getChangeAddress) (pure <<< Just)
       =<< asksConstraints Constraints._changeAddress
 
+  getAddressesAndUtxos
+    :: Maybe Wallet -> BalanceTxM UtxoMap
+  getAddressesAndUtxos mWallet = do
+    srcAddrs <-
+      asksConstraints Constraints._srcAddresses
+        >>= maybe (liftQueryM QueryM.getWalletAddresses) pure
+
+    utxos <- liftEitherQueryM
+      case mWallet of
+        Just (Eternl _) -> do
+          walletAddresses <- QueryM.getWalletAddresses
+          let
+            { yes: inWalletAddresses, no: outWalletAddresses } = partition
+              (\addr -> elem addr walletAddresses)
+              srcAddrs
+          inWalletUtxos <- repeatN 5 $ getWalletUtxosAt inWalletAddresses
+          outWalletUtxos <- parallelUtxosAt outWalletAddresses
+          pure $ Map.union <$> inWalletUtxos <*> outWalletUtxos
+        _ -> parallelUtxosAt srcAddrs
+    if Map.isEmpty utxos then
+      do
+        logInfo' "empty utxos, attempting again"
+        getAddressesAndUtxos mWallet
+    else pure utxos
+
+  getWalletUtxosAt :: Array Address -> QueryM (Either BalanceTxError UtxoMap)
+  getWalletUtxosAt inWalletAddresses = do
+    filteredWalletUtxos <-
+      maybe Map.empty
+        ( Map.filter
+            (\txout -> elem (_.address (unwrap txout)) inWalletAddresses)
+        ) <$> getWalletUtxos
+    utxosAtWalletAddresses <- parallelUtxosAt inWalletAddresses
+    pure $ case Map.isSubmap filteredWalletUtxos <$> utxosAtWalletAddresses of
+      Left e -> Left e
+      Right false -> Left CouldNotGetUtxos
+      Right true -> Right $ filteredWalletUtxos
+
+  repeatN
+    :: forall (m :: Type -> Type) (b :: Type) (c :: Type)
+     . Monad m
+    => MonadLogger m
+    => MonadAff m
+    => Int
+    -> m (Either b c)
+    -> m (Either b c)
+  repeatN attempts action =
+    if attempts <= 0 then action
+    else do
+      result <- action
+      case result of
+        Left _ -> do
+          logInfo' $ "waiting repeatN: " <> show attempts
```

Comment:
I would like to preserv logs of the wait but I need to change them to `Trace` or `Debug`

I would refactor `repeatN` function anyway.